File: rock/env_vars.py
import json
import os
import sys
from collections.abc import Callable
from pathlib import Path
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    ROCK_LOGGING_PATH: str | None = None
    ROCK_LOGGING_FILE_NAME: str | None = None
    ROCK_LOGGING_LEVEL: str | None = None
    ROCK_SERVICE_STATUS_DIR: str | None = None
    ROCK_SCHEDULER_STATUS_DIR: str | None = None
    ROCK_CONFIG: str | None = None
    ROCK_CONFIG_DIR_NAME: str | None = None
    ROCK_BASE_URL: str | None = "http://localhost:8080"
    ROCK_WORKER_ROCKLET_PORT: int | None = None
    ROCK_SANDBOX_STARTUP_TIMEOUT_SECONDS: int = 180
    ROCK_CODE_SANDBOX_BASE_URL: str | None = None
    ROCK_ENVHUB_BASE_URL: str | None = "http://localhost:8081"
    ROCK_ENVHUB_DEFAULT_DOCKER_IMAGE: str | None = "python:3.11"
    ROCK_ENVHUB_DB_URL: str | None = f"sqlite:///{Path.home() / '.rock' / 'rock_envs.db'}"
    ROCK_DEFAULT_AUTO_CLEAR_TIME_MINUTES: int = 60 * 6  # 6 hours
    ROCK_RAY_NAMESPACE: str | None = "xrl-sandbox"
    ROCK_SANDBOX_EXPIRE_TIME_KEY: str | None = "expire_time"
    ROCK_SANDBOX_AUTO_CLEAR_TIME_KEY: str | None = "auto_clear_time"
    ROCK_TIME_ZONE: str = "Asia/Shanghai"

    # Scheduler
    ROCK_DOCUUM_INSTALL_URL: str | None = None

    # OSS Config
    ROCK_OSS_ENABLE: bool = False
    ROCK_OSS_BUCKET_ENDPOINT: str | None = None
    ROCK_OSS_BUCKET_NAME: str | None = None
    ROCK_OSS_BUCKET_REGION: str | None = None

    ROCK_PIP_INDEX_URL: str | None = "https://mirrors.aliyun.com/pypi/simple/"
    ROCK_MONITOR_ENABLE: bool = False
    ROCK_PROJECT_ROOT: str | None = None
    ROCK_WORKER_ENV_TYPE: str | None = "local"
    ROCK_PYTHON_ENV_PATH: str | None = None
    ROCK_ADMIN_ENV: str | None = "dev"
    ROCK_ADMIN_ROLE: str | None = "write"
    ROCK_CLI_LOAD_PATHS: str = str(Path(__file__).parent / "cli" / "command")
    ROCK_CLI_DEFAULT_CONFIG_PATH: str

    # Model Service Config
    ROCK_MODEL_SERVICE_DATA_DIR: str
    ROCK_MODEL_SERVICE_TRAJ_APPEND_MODE: bool | None = None

    # RuntimeEnv
    ROCK_RTENV_PYTHON_V31114_INSTALL_CMD: str
    ROCK_RTENV_PYTHON_V31212_INSTALL_CMD: str
    ROCK_RTENV_NODE_V22180_INSTALL_CMD: str

    # Agentic
    ROCK_AGENT_PRE_INIT_BASH_CMD_LIST: list[str] = []

    ROCK_AGENT_IFLOW_CLI_INSTALL_CMD: str

    ROCK_MODEL_SERVICE_INSTALL_CMD: str


environment_variables: dict[str, Callable[[], Any]] = {
    "ROCK_LOGGING_PATH": lambda: os.getenv("ROCK_LOGGING_PATH"),
    "ROCK_LOGGING_FILE_NAME": lambda: os.getenv("ROCK_LOGGING_FILE_NAME", "rocklet.log"),
    "ROCK_LOGGING_LEVEL": lambda: os.getenv("ROCK_LOGGING_LEVEL", "INFO"),
    "ROCK_SERVICE_STATUS_DIR": lambda: os.getenv("ROCK_SERVICE_STATUS_DIR", "/data/service_status"),
    "ROCK_SCHEDULER_STATUS_DIR": lambda: os.getenv("ROCK_SCHEDULER_STATUS_DIR", "/data/scheduler_status"),
    "ROCK_CONFIG": lambda: os.getenv("ROCK_CONFIG"),
    "ROCK_CONFIG_DIR_NAME": lambda: os.getenv("ROCK_CONFIG_DIR_NAME", "rock-conf"),
    "ROCK_BASE_URL": lambda: os.getenv("ROCK_BASE_URL", "http://localhost:8080"),
    "ROCK_WORKER_ROCKLET_PORT": lambda: int(val) if (val := os.getenv("ROCK_WORKER_ROCKLET_PORT")) else None,
    "ROCK_SANDBOX_STARTUP_TIMEOUT_SECONDS": lambda: int(os.getenv("ROCK_SANDBOX_STARTUP_TIMEOUT_SECONDS", "180")),
    "ROCK_CODE_SANDBOX_BASE_URL": lambda: os.getenv("ROCK_CODE_SANDBOX_BASE_URL", ""),
    "ROCK_ENVHUB_BASE_URL": lambda: os.getenv("ROCK_ENVHUB_BASE_URL", "http://localhost:8081"),
    "ROCK_ENVHUB_DEFAULT_DOCKER_IMAGE": lambda: os.getenv("ROCK_ENVHUB_DEFAULT_DOCKER_IMAGE", "python:3.11"),
    "ROCK_ENVHUB_DB_URL": lambda: os.getenv(
        "ROCK_ENVHUB_DB_URL", f"sqlite:///{Path.home() / '.rock' / 'rock_envs.db'}"
    ),
    "ROCK_DEFAULT_AUTO_CLEAR_TIME_MINUTES": lambda: int(os.getenv("ROCK_DEFAULT_AUTO_CLEAR_TIME_MINUTES", "360")),
    "ROCK_RAY_NAMESPACE": lambda: os.getenv("ROCK_RAY_NAMESPACE", "xrl-sandbox"),
    "ROCK_SANDBOX_EXPIRE_TIME_KEY": lambda: os.getenv("ROCK_SANDBOX_EXPIRE_TIME_KEY", "expire_time"),
    "ROCK_SANDBOX_AUTO_CLEAR_TIME_KEY": lambda: os.getenv("ROCK_SANDBOX_AUTO_CLEAR_TIME_KEY", "auto_clear_time"),
    "ROCK_OSS_ENABLE": lambda: os.getenv("ROCK_OSS_ENABLE", "false").lower() == "true",
    "ROCK_OSS_BUCKET_ENDPOINT": lambda: os.getenv("ROCK_OSS_BUCKET_ENDPOINT"),
    "ROCK_OSS_BUCKET_NAME": lambda: os.getenv("ROCK_OSS_BUCKET_NAME"),
    "ROCK_OSS_BUCKET_REGION": lambda: os.getenv("ROCK_OSS_BUCKET_REGION"),
    "ROCK_PIP_INDEX_URL": lambda: os.getenv("ROCK_PIP_INDEX_URL", "https://mirrors.aliyun.com/pypi/simple/"),
    "ROCK_MONITOR_ENABLE": lambda: os.getenv("ROCK_MONITOR_ENABLE", "false").lower() == "true",
    "ROCK_PROJECT_ROOT": lambda: os.getenv("ROCK_PROJECT_ROOT", str(Path(__file__).resolve().parents[1])),
    "ROCK_WORKER_ENV_TYPE": lambda: os.getenv("ROCK_WORKER_ENV_TYPE", "local"),
    "ROCK_PYTHON_ENV_PATH": lambda: os.getenv("ROCK_PYTHON_ENV_PATH", sys.base_prefix),
    "ROCK_ADMIN_ENV": lambda: os.getenv("ROCK_ADMIN_ENV", "dev"),
    "ROCK_ADMIN_ROLE": lambda: os.getenv("ROCK_ADMIN_ROLE", "write"),
    "ROCK_CLI_LOAD_PATHS": lambda: os.getenv("ROCK_CLI_LOAD_PATHS", str(Path(__file__).parent / "cli" / "command")),
    "ROCK_CLI_DEFAULT_CONFIG_PATH": lambda: os.getenv(
        "ROCK_CLI_DEFAULT_CONFIG_PATH", Path.home() / ".rock" / "config.ini"
    ),
    "ROCK_MODEL_SERVICE_DATA_DIR": lambda: os.getenv("ROCK_MODEL_SERVICE_DATA_DIR", "/data/logs"),
    "ROCK_MODEL_SERVICE_TRAJ_APPEND_MODE": lambda: os.getenv("ROCK_MODEL_SERVICE_TRAJ_APPEND_MODE", "false").lower()
    == "true",
    "ROCK_RTENV_PYTHON_V31114_INSTALL_CMD": lambda: os.getenv(
        "ROCK_RTENV_PYTHON_V31114_INSTALL_CMD",
        "[ -f cpython31114.tar.gz ] && rm cpython31114.tar.gz; [ -d python ] && rm -rf python; wget -q -O cpython31114.tar.gz https://github.com/astral-sh/python-build-standalone/releases/download/20251120/cpython-3.11.14+20251120-x86_64-unknown-linux-gnu-install_only.tar.gz && tar -xzf cpython31114.tar.gz && mv python runtime-env",
    ),
    "ROCK_RTENV_PYTHON_V31212_INSTALL_CMD": lambda: os.getenv(
        "ROCK_RTENV_PYTHON_V31212_INSTALL_CMD",
        "[ -f cpython-31212.tar.gz ] && rm cpython-31212.tar.gz; [ -d python ] && rm -rf python; wget -q -O cpython-31212.tar.gz https://github.com/astral-sh/python-build-standalone/releases/download/20251217/cpython-3.12.12+20251217-x86_64-unknown-linux-gnu-install_only.tar.gz && tar -xzf cpython-31212.tar.gz && mv python runtime-env",
    ),
    "ROCK_RTENV_NODE_V22180_INSTALL_CMD": lambda: os.getenv(
        "ROCK_RTENV_NODE_V22180_INSTALL_CMD",
        "[ -f node.tar.xz ] && rm node.tar.xz; [ -d node ] && rm -rf node; wget -q -O node.tar.xz --tries=10 --waitretry=2 https://nodejs.org/dist/v22.18.0/node-v22.18.0-linux-x64.tar.xz && tar -xf node.tar.xz && mv node-v22.18.0-linux-x64 runtime-env",
    ),
    "ROCK_AGENT_PRE_INIT_BASH_CMD_LIST": lambda: json.loads(os.getenv("ROCK_AGENT_PRE_INIT_BASH_CMD_LIST", "[]")),
    "ROCK_AGENT_IFLOW_CLI_INSTALL_CMD": lambda: os.getenv(
        "ROCK_AGENT_IFLOW_CLI_INSTALL_CMD",
        "npm i -g @iflow-ai/iflow-cli@latest",
    ),
    "ROCK_MODEL_SERVICE_INSTALL_CMD": lambda: os.getenv(
        "ROCK_MODEL_SERVICE_INSTALL_CMD",
        "pip install rl_rock[model-service]",
    ),
    "ROCK_TIME_ZONE": lambda: os.getenv("ROCK_TIME_ZONE", "Asia/Shanghai"),
    "ROCK_DOCUUM_INSTALL_URL": lambda: os.getenv(
        "ROCK_DOCUUM_INSTALL_URL", "https://raw.githubusercontent.com/stepchowfun/docuum/main/install.sh"
    ),
}


def __getattr__(name: str):
    if name in environment_variables:
        return environment_variables[name]()
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")


def is_set(name: str):
    """Check if an environment variable is explicitly set."""
    if name in environment_variables:
        return name in os.environ
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")

File: rock/ts-sdk/src/env_vars.ts
/**
 * Environment variables configuration
 */

import { getEnv, isEnvSet } from './utils/system.js';
import { homedir } from 'os';
import { join } from 'path';

/**
 * Environment variable definitions
 */
export const envVars = {
  // Logging
  get ROCK_LOGGING_PATH(): string | undefined {
    return getEnv('ROCK_LOGGING_PATH');
  },

  get ROCK_LOGGING_FILE_NAME(): string {
    return getEnv('ROCK_LOGGING_FILE_NAME', 'rocklet.log')!;
  },

  get ROCK_LOGGING_LEVEL(): string {
    return getEnv('ROCK_LOGGING_LEVEL', 'INFO')!;
  },

  // Service
  get ROCK_SERVICE_STATUS_DIR(): string {
    return getEnv('ROCK_SERVICE_STATUS_DIR', '/data/service_status')!;
  },

  get ROCK_SCHEDULER_STATUS_DIR(): string {
    return getEnv('ROCK_SCHEDULER_STATUS_DIR', '/data/scheduler_status')!;
  },

  // Config
  get ROCK_CONFIG(): string | undefined {
    return getEnv('ROCK_CONFIG');
  },

  get ROCK_CONFIG_DIR_NAME(): string {
    return getEnv('ROCK_CONFIG_DIR_NAME', 'rock-conf')!;
  },

  // Base URLs
  get ROCK_BASE_URL(): string {
    return getEnv('ROCK_BASE_URL', 'http://localhost:8080')!;
  },

  get ROCK_WORKER_ROCKLET_PORT(): number | undefined {
    const val = getEnv('ROCK_WORKER_ROCKLET_PORT');
    return val ? parseInt(val, 10) : undefined;
  },

  get ROCK_SANDBOX_STARTUP_TIMEOUT_SECONDS(): number {
    return parseInt(getEnv('ROCK_SANDBOX_STARTUP_TIMEOUT_SECONDS', '180')!, 10);
  },

  get ROCK_CODE_SANDBOX_BASE_URL(): string {
    return getEnv('ROCK_CODE_SANDBOX_BASE_URL', '')!;
  },

  // EnvHub
  get ROCK_ENVHUB_BASE_URL(): string {
    return getEnv('ROCK_ENVHUB_BASE_URL', 'http://localhost:8081')!;
  },

  get ROCK_ENVHUB_DEFAULT_DOCKER_IMAGE(): string {
    return getEnv('ROCK_ENVHUB_DEFAULT_DOCKER_IMAGE', 'python:3.11')!;
  },

  get ROCK_ENVHUB_DB_URL(): string {
    return getEnv(
      'ROCK_ENVHUB_DB_URL',
      `sqlite:///${join(homedir(), '.rock', 'rock_envs.db')}`
    )!;
  },

  // Auto clear
  get ROCK_DEFAULT_AUTO_CLEAR_TIME_MINUTES(): number {
    return parseInt(getEnv('ROCK_DEFAULT_AUTO_CLEAR_TIME_MINUTES', '360')!, 10);
  },

  // Ray
  get ROCK_RAY_NAMESPACE(): string {
    return getEnv('ROCK_RAY_NAMESPACE', 'xrl-sandbox')!;
  },

  get ROCK_SANDBOX_EXPIRE_TIME_KEY(): string {
    return getEnv('ROCK_SANDBOX_EXPIRE_TIME_KEY', 'expire_time')!;
  },

  get ROCK_SANDBOX_AUTO_CLEAR_TIME_KEY(): string {
    return getEnv('ROCK_SANDBOX_AUTO_CLEAR_TIME_KEY', 'auto_clear_time')!;
  },

  // Timezone
  get ROCK_TIME_ZONE(): string {
    return getEnv('ROCK_TIME_ZONE', 'Asia/Shanghai')!;
  },

  // OSS
  get ROCK_OSS_ENABLE(): boolean {
    return getEnv('ROCK_OSS_ENABLE', 'false')?.toLowerCase() === 'true';
  },

  get ROCK_OSS_BUCKET_ENDPOINT(): string | undefined {
    return getEnv('ROCK_OSS_BUCKET_ENDPOINT');
  },

  get ROCK_OSS_BUCKET_NAME(): string | undefined {
    return getEnv('ROCK_OSS_BUCKET_NAME');
  },

  get ROCK_OSS_BUCKET_REGION(): string | undefined {
    return getEnv('ROCK_OSS_BUCKET_REGION');
  },

  get ROCK_OSS_TIMEOUT(): number {
    return parseInt(getEnv('ROCK_OSS_TIMEOUT', '300000')!, 10); // Default: 5 minutes
  },

  // Pip
  get ROCK_PIP_INDEX_URL(): string {
    return getEnv('ROCK_PIP_INDEX_URL', 'https://pypi.org/simple/')!;
  },

  // Monitor
  get ROCK_MONITOR_ENABLE(): boolean {
    return getEnv('ROCK_MONITOR_ENABLE', 'false')?.toLowerCase() === 'true';
  },

  // Project
  get ROCK_PROJECT_ROOT(): string {
    return getEnv('ROCK_PROJECT_ROOT', process.cwd())!;
  },

  get ROCK_WORKER_ENV_TYPE(): string {
    return getEnv('ROCK_WORKER_ENV_TYPE', 'local')!;
  },

  get ROCK_PYTHON_ENV_PATH(): string {
    return getEnv('ROCK_PYTHON_ENV_PATH', process.cwd())!;
  },

  // Admin
  get ROCK_ADMIN_ENV(): string {
    return getEnv('ROCK_ADMIN_ENV', 'dev')!;
  },

  get ROCK_ADMIN_ROLE(): string {
    return getEnv('ROCK_ADMIN_ROLE', 'write')!;
  },

  // CLI
  get ROCK_CLI_LOAD_PATHS(): string {
    return getEnv('ROCK_CLI_LOAD_PATHS', '')!;
  },

  get ROCK_CLI_DEFAULT_CONFIG_PATH(): string {
    return getEnv('ROCK_CLI_DEFAULT_CONFIG_PATH', join(homedir(), '.rock', 'config.ini'))!;
  },

  // Model Service
  get ROCK_MODEL_SERVICE_DATA_DIR(): string {
    return getEnv('ROCK_MODEL_SERVICE_DATA_DIR', '/data/logs')!;
  },

  get ROCK_MODEL_SERVICE_TRAJ_APPEND_MODE(): boolean {
    return getEnv('ROCK_MODEL_SERVICE_TRAJ_APPEND_MODE', 'false')?.toLowerCase() === 'true';
  },

  // RuntimeEnv
  get ROCK_RTENV_PYTHON_V31114_INSTALL_CMD(): string {
    return getEnv(
      'ROCK_RTENV_PYTHON_V31114_INSTALL_CMD',
      '[ -f cpython31114.tar.gz ] && rm cpython31114.tar.gz; [ -d python ] && rm -rf python; wget -q -O cpython31114.tar.gz https://github.com/astral-sh/python-build-standalone/releases/download/20251120/cpython-3.11.14+20251120-x86_64-unknown-linux-gnu-install_only.tar.gz && tar -xzf cpython31114.tar.gz && mv python runtime-env'
    )!;
  },

  get ROCK_RTENV_PYTHON_V31212_INSTALL_CMD(): string {
    return getEnv(
      'ROCK_RTENV_PYTHON_V31212_INSTALL_CMD',
      '[ -f cpython-3.12.12.tar.gz ] && rm cpython-3.12.12.tar.gz; [ -d python ] && rm -rf python; wget -q -O cpython-3.12.12.tar.gz https://github.com/astral-sh/python-build-standalone/releases/download/20251217/cpython-3.12.12+20251217-x86_64-unknown-linux-gnu-install_only.tar.gz && tar -xzf cpython-3.12.12.tar.gz && mv python runtime-env'
    )!;
  },

  get ROCK_RTENV_NODE_V22180_INSTALL_CMD(): string {
    return getEnv(
      'ROCK_RTENV_NODE_V22180_INSTALL_CMD',
      '[ -f node.tar.xz ] && rm node.tar.xz; [ -d node ] && rm -rf node; wget -q -O node.tar.xz --tries=10 --waitretry=2 https://nodejs.org/dist/v22.18.0/node-v22.18.0-linux-x64.tar.xz && tar -xf node.tar.xz && mv node-v22.18.0-linux-x64 runtime-env'
    )!;
  },

  // Agent
  get ROCK_AGENT_PRE_INIT_BASH_CMD_LIST(): Array<{ command: string; timeoutSeconds: number }> {
    const val = getEnv('ROCK_AGENT_PRE_INIT_BASH_CMD_LIST', '[]');
    try {
      return JSON.parse(val!);
    } catch {
      return [];
    }
  },

  get ROCK_AGENT_IFLOW_CLI_INSTALL_CMD(): string {
    return getEnv('ROCK_AGENT_IFLOW_CLI_INSTALL_CMD', 'npm i -g @iflow-ai/iflow-cli@latest')!;
  },

  get ROCK_MODEL_SERVICE_INSTALL_CMD(): string {
    return getEnv('ROCK_MODEL_SERVICE_INSTALL_CMD', 'pip install rl_rock[model-service]')!;
  },

  // Doccuum
  get ROCK_DOCUUM_INSTALL_URL(): string {
    return getEnv(
      'ROCK_DOCUUM_INSTALL_URL',
      'https://raw.githubusercontent.com/stepchowfun/docuum/main/install.sh'
    )!;
  },

  // ========== Sandbox Defaults ==========
  // Sandbox configuration defaults - allow users to override via environment variables

  get ROCK_DEFAULT_IMAGE(): string {
    return getEnv('ROCK_DEFAULT_IMAGE', 'python:3.11')!;
  },

  get ROCK_DEFAULT_MEMORY(): string {
    return getEnv('ROCK_DEFAULT_MEMORY', '8g')!;
  },

  get ROCK_DEFAULT_CPUS(): number {
    return parseFloat(getEnv('ROCK_DEFAULT_CPUS', '2')!);
  },

  get ROCK_DEFAULT_CLUSTER(): string {
    return getEnv('ROCK_DEFAULT_CLUSTER', 'zb')!;
  },

  get ROCK_DEFAULT_AUTO_CLEAR_SECONDS(): number {
    return parseInt(getEnv('ROCK_DEFAULT_AUTO_CLEAR_SECONDS', '300')!, 10);
  },

  // ========== SandboxGroup Defaults ==========

  get ROCK_DEFAULT_GROUP_SIZE(): number {
    return parseInt(getEnv('ROCK_DEFAULT_GROUP_SIZE', '2')!, 10);
  },

  get ROCK_DEFAULT_START_CONCURRENCY(): number {
    return parseInt(getEnv('ROCK_DEFAULT_START_CONCURRENCY', '2')!, 10);
  },

  get ROCK_DEFAULT_START_RETRY_TIMES(): number {
    return parseInt(getEnv('ROCK_DEFAULT_START_RETRY_TIMES', '3')!, 10);
  },

  // ========== Client Timeouts (in seconds) ==========

  get ROCK_DEFAULT_ARUN_TIMEOUT(): number {
    return parseInt(getEnv('ROCK_DEFAULT_ARUN_TIMEOUT', '300')!, 10);
  },

  get ROCK_DEFAULT_NOHUP_WAIT_TIMEOUT(): number {
    return parseInt(getEnv('ROCK_DEFAULT_NOHUP_WAIT_TIMEOUT', '300')!, 10);
  },

  get ROCK_DEFAULT_NOHUP_WAIT_INTERVAL(): number {
    return parseInt(getEnv('ROCK_DEFAULT_NOHUP_WAIT_INTERVAL', '10')!, 10);
  },

  get ROCK_DEFAULT_STATUS_CHECK_INTERVAL(): number {
    return parseInt(getEnv('ROCK_DEFAULT_STATUS_CHECK_INTERVAL', '3')!, 10);
  },
};

/**
 * Check if an environment variable is explicitly set
 */
export function isSet(name: string): boolean {
  return isEnvSet(name);
}
